content/articles/memoization-and-dp-in-python-beginners-guide/index.md

```diff
@@ -0,0 +1,122 @@
+our function is going to look like this:
+```python
+def knapsack(W,weights, values, n):
+
+```
+where W is the remaining capacity of the knapsack, weights is the array of weights and values is the array of values our items are worth. n is the length of both arrays.
+A recursive function must have a base case. Our base case here is going to be when the knapsack is full(when the remaining available weight we are allowed to add is equal to 0) or when there are no more items on the list
+```python
+    if n==0 or W==0:
+        return 0
+```
+next, we are going to decide whether to include an item or ignore it.
+To do this we have to first check whether the weight of the item is bigger than the remaining capacity.
+If the weight of the current item is less than the allowed capacity, we get the max of either ignoring it or adding it to the bag.
+```python
+    if weights[n-1]<=W:
+        return max(values[n-1]+knapsack(W-weights[n-1], weights, values, n-1), knapsack(W, weights, values, n-1))
+```
+If it is, we ignore it and recursively call self with the same values minus this current element.
+```python
+    else:
+        return knapsack(W, weights, values, n-1)
+```
+
+
+Whether we choose to include it or ignore it, we have to 'pop' the current element from the array, that is we make the next recursive call without it.
+This is easy to achieve because we are passing the length of the array, and we are using indices for comparison. All we have to do is decrement the size of the array and in the next call, we are going to be using the next item.
+This recursive solution takes exponential time, let's see how we can optimize it to reduce the time complexity.
+
+### how can we use memoization in our function? <a name='solution'></a>
+As you can see, there are two values that our recursive function depends on, the remaining capacity that we are allowed to add to the knapsack and the number of items.
+We can save the results of the maximum weight when we consider each item so that we don't have to recalculate it every time we encounter it.
+Before executing a calculation we can check whether we had already saved the result of such a calculation.
+If it exists, we return it instead of recalculating it. We can choose to save them to a regular list or a dictionary. A dictionary has faster lookups so let's use it.
+```python
+
+    memodict = {} #initialize empty dict
+    def knapsack(W,weights, values, n):
+        if n==0 or W==0:
+            return 0
+        key = str(n)+str(W) #create a string key for each calculation
+        val = memodict.get(key)
+        if val is not None:
+            return val
+        if weights[n-1]<=W:
+            memodict[key] = max(values[n-1]+knapsack(W-weights[n-1], weights, values, n-1), knapsack(W, weights, values, n-1))
+        elif weights[n-1]>W:
+            memodict[key] = knapsack(W, weights, values, n-1)
+
+        return memodict[key]
+
+```
+#Complexity Analysis
```

Comment:
This is a huge topic.

I can't believe you just covered them in 2 lines. How did you derive this time or space complexity. Explain step by step.

Before all, first explain what complexity analysis is.